### PR TITLE
Add river-ocean two-way coupling feature 

### PR DESCRIPTION
**Pull request by fdongyu**
This development implements the river-ocean two-way coupling. This feature is turned off by default in our development and can be turned on by setting use_lnd_rof_two_way=.true. in user_nl_mosart. All e3sm_developer test suites passed.

This new feature enables two-way coupling between MPAS-O and MOSART. The water fluxes from MOSART to MPAS-O are treated the same as previously. The MPAS-O simulated sea surface height (ssh) time series are remapped to the MOSART outlet cells and are used as the downstream, dynamic water stage boundary of the specified outlets, which will further propagate to the upstream rivers. The two-way coupling is only configured for MOSART outlets specified by ocn_rof_coupling_ID=1 in the MOSART parameter input file (frivinp_rtm), while the other outlets in the model domain (ocn_rof_coupling_ID=0) are still one-way coupled. Another parameter in frivinp_rtm, vdatum_conversion, is also required to normalize the MPAS-O modeled ssh to MOSART water depth based on local river geometry.

### CODE REVIEW

**Comment by liho745** [comment]
The title/description of this PR can be revised to be more accurate. Since the ocean is only invoked as the boundary condition to the river, what'd done is here not strictly "two-way" coupling, but rather something like "effects of sea surface height changes on river dynamics".

File: components/mosart/cime_config/testdefs/testmods_dirs/mosart/rof_ocn_2way/user_nl_mosart
Comment (liho745) [inline components/mosart/cime_config/testdefs/testmods_dirs/mosart/rof_ocn_2way/user_nl_mosart:2]: The value of routingmethod can be revised in RtmMod.F90. So instead of using two flags to turn on this feature, I'd suggest to use only one flag: use_ocn_rof_two_way. You can add some lines in RtmMod to check whether use_ocn_rof_two_way is true, and if true, make routingmethod equal to DW.

Comment (fdongyu) [reply components/mosart/cime_config/testdefs/testmods_dirs/mosart/rof_ocn_2way/user_nl_mosart]: Thanks. The modification was made in RtmMod.F90 to force routingmethod=DW when the use_ocn_rof_two_way flag is on.

**Comment by fdongyu** [comment]
> The title/description of this PR can be revised to be more accurate. Since the ocean is only invoked as the boundary condition to the river, what'd done is here not strictly "two-way" coupling, but rather something like "effects of sea surface height changes on river dynamics". In my understanding, a real "two-way" coupling means both MPAS and MOSART run together and exchange fluxes in each coupling time step.

I think this confusion may be due to the aforementioned manuscript, in which MPAS-O is not running together with MOSART. This manuscript is only used to demonstrate the advantage of creating a downstream boundary in MOSART and does not strictly implement the two-way coupling because the MPAS-O simulated sea surface height is not accurate enough. This branch, on the other hand, enables MOSART and MPAS-O running simultaneously and exchanging information through the coupler, i.e., at each coupling time step, MOSART uses the updated sea surface height from MPAS-O. It may be appropriate to still name this PR as river-ocean two-way coupling since a majority of code modifications are made in the coupler.

**Comment by liho745** [comment]
@fdongyu Thanks for the explanation. Could you further revise the PR description to 1) revise the part mentioning routingmethod since you've made the change in RtmMod.F90, 2) add a bit more details on what you mean by two-way coupling, e.g., something like "Adding two-way coupling between MPAS-O and MOSART. The water fluxes from MOSART to MPAS-O are still treated the same as previously. The MPAS-O simulated sea surface height (ssh) time series are remapped to the MOSART outlet cells and is used as the downstream, dynamic water stage boundary of the specified outlets, which will further propagate to the upstream rivers...", 3) perhaps you need to mention the ICoM project or your manuscript here?

**Comment by fdongyu** [comment]
> @fdongyu Thanks for the explanation. Could you further revise the PR description to 1) revise the part mentioning routingmethod since you've made the change in RtmMod.F90, 2) add a bit more details on what you mean by two-way coupling, e.g., something like "Adding two-way coupling between MPAS-O and MOSART. The water fluxes from MOSART to MPAS-O are still treated the same as previously. The MPAS-O simulated sea surface height (ssh) time series are remapped to the MOSART outlet cells and is used as the downstream, dynamic water stage boundary of the specified outlets, which will further propagate to the upstream rivers...", 3) perhaps you don't need to mention the ICoM project or your manuscript here?

Thanks @liho745, the description was edited accordingly.

**Review by bishtgautam — CHANGES_REQUESTED** [review]
@fdongyu, Use a namelist variable in the driver instead of the MOSART to determine if the two-way coupling is turned on. If the two-way coupling is on:
-  Send `ssh` from the ocean to the coupler
-  In MOSART, use `index_x2r_So_ssh > 0` to figure out if the two-way coupling is turned on.

@donghuix made a similar change in PR #5017 in https://github.com/E3SM-Project/E3SM/pull/5017/commits/715e501d87d95a284221986007cb8a96bb541746

**Comment by fdongyu** [comment]
> @fdongyu, Use a namelist variable in the driver instead of the MOSART to determine if the two-way coupling is turned on. If the two-way coupling is on:
> 
> * Send `ssh` from the ocean to the coupler
> * In MOSART, use `index_x2r_So_ssh > 0` to figure out if the two-way coupling is turned on.
> 
> @donghuix made a similar change in PR #5017 in [715e501](https://github.com/E3SM-Project/E3SM/commit/715e501d87d95a284221986007cb8a96bb541746)

Thanks, @bishtgautam.  Modifications were made accordingly and all tests passed.

**Comment by bishtgautam** [comment]
@fdongyu Thanks for making edits. I think the many conflicts of this PR are due to the merge of @donghuix's recent PR. Can you now try to rebase your branch to avoid conflicts?

**Comment by fdongyu** [comment]
> @fdongyu Thanks for making edits. I think the many conflicts of this PR are due to the merge of @donghuix's recent PR. Can you now try to rebase your branch to avoid conflicts?

This branch is now rebased to Donghui's recent PR. The conflicts are resolved and all developer tests passed too.

**Review by jonbob — COMMENTED** [review]
Other than the question about the mapping file, I did check out this code and successfully ran a case using GPMPAS-JRA and ne30pg2_r05_EC30to60E2r2 with both ocn_rof_two_way false and true. As soon as that question is resolved I'm happy to approve

File: cime_config/config_grids.xml
Comment (fdongyu) [reply cime_config/config_grids.xml:4158]: @jonbob Thanks for the question. The mapping file remaps MPAS-O ssh to MOSART ocean cells and outlet cells. The ssh mapped to the outlet cells will be used as the downstream boundary of MOSART. Because there are usually no overlaps between the outlets and the ocean mesh, I use the ssh of the two nearest neighboring ocean cells for an outlet to calculate the mapped ssh instead of using the bilinear interpolation. The weight for each neighbor is the inverse of distance between the outlet and the ocean cell. This is equivalent to inverse distance weighting interpolation and the number of neighbored ocean cells can be arbitrary. I found 2 is adequate for this mesh. I agree with you that we may need to change the remapping method when the high-resolution unstructured mesh is available. The key is to make sure which ocean cells are mapped to the river outlets since they are not supposed to overlap.

Comment (rljacob) [reply cime_config/config_grids.xml:4158]: Is this mapping file made with a custom program like the r2o one?  Or did you use ncremap/Tempest/ESMF ?

Comment (fdongyu) [reply cime_config/config_grids.xml:4158]: @rljacob, yes, the mapping file has to be made with a custom program. The MPAS-O ssh needs to be mapped to MOSART outlets. But in addition to the ocean outlets along the coastline, MOSART has dummy outlets on the land, in which discharge will be routed to the nearest ocean outlets. These land outlets have to be excluded from the remapping. So I use a custom remapping program that defines which outlets are included in a shapefile.

Comment (jonbob) [reply cime_config/config_grids.xml:4158]: @fdongyu - is that custom program documented and available for use?

Comment (fdongyu) [reply cime_config/config_grids.xml:4158]: @jonbob The program is a Python script and has only been shared with ICoM team members. If it has to come with this PR, please let me know the best place to publish it (my own git repo or zenodo?). Thanks!

Comment (jonbob) [reply cime_config/config_grids.xml:4158]: @rljacob - what do you think? @fdongyu, I'm a bit curious how this is easier than mapping the ocean SSH to all the corresponding points on the global mosart grid with a more typical mapping, and then just using the outlet cell neighbors' value for SSH?

Comment (fdongyu) [reply cime_config/config_grids.xml:4158]: @jonbob , thanks for the suggestion! I will do it that way. There would not be too much difference except that I will let each outlet search the nearest ocean ssh. Could you help me generate a mapping file for this example (EC30to60E2r2 to r05)? I tried the NCO remapping but got some errors.

Comment (jonbob) [reply cime_config/config_grids.xml:4158]: @fdongyu - I made a bilinear map, which we typically use for remapping states, for this and put it on lcrc at:
````
/lcrc/group/e3sm/data/inputdata/cpl/cpl6/map_EC30to60E2r2_to_r05_bilin.220727.nc
````
There is no masking done so you can hopefully get SSH anywhere the ocean exists on the r05 grid. But please let me know if you'd like to try something else

Comment (fdongyu) [reply cime_config/config_grids.xml:4158]: @jonbob , thanks very much! I wonder if it is possible to map the MPAS-O SSH to all ocean cells. The ocean cells are identified in the land domain file. This may include extrapolations for the r05 cells that are identified as ocean by the land domain file but do not have overlap with the EC30to60E2r2 ocean grid. 
Otherwise, it may be difficult to tell a MOSART outlet where the nearest SSH is, if its nearest ocean cell does not have a remapped SSH. Please let me know if my description is not clear.

Comment (jonbob) [reply cime_config/config_grids.xml:4158]: Thanks for testing that, @fdongyu. And I think I understand what you're asking. It is tricky because remapping files in general apply only to the overlap between grids, and as you noticed the mpas-o grid may not correspond exactly with the inverse of the lnd domain in the r05 mosart file. Let me try a couple of things and see if I can come up with something else.

Comment (jonbob) [reply cime_config/config_grids.xml:4158]: @fdongyu  - Well, I have one other file for you to try -- but it's just a nearest-neighbor mapping created with ncremap:
````
/lcrc/group/e3sm/data/inputdata/cpl/cpl6/map_EC30to60E2r2_to_r05_neareststod.220728.nc
````
I'll ask someone else for any ideas, and if we don't come up with anything else your tool maybe the best answer. Thanks for your patience

Comment (fdongyu) [reply cime_config/config_grids.xml:4158]: Thanks, @jonbob. I think this nearest-neighbor mapping works in a similar way as my approach, except that all land cells get a remapped ssh (see the remapped ssh below). I think this is fine since the ssh of the land cells will not be used. Please let me know if you have an updated solution. If not, I am fine with proceeding with this mapping file.  
remapped_ssh

Comment (rljacob) [reply cime_config/config_grids.xml:4158]: If it turns out we're going to use that python script for making maps, @fdongyu you can create your own repository with it and we can add it as an external module to mosart.   Or you can add it directly to the E3SM code (in components/mosart/utils) and put any DOI or Copyright info in a comment at the top of the file.

Comment (ambrad) [reply cime_config/config_grids.xml:4158]: @rljacob as another possibility, what do you think of https://github.com/E3SM-Project/E3SM-init?

Comment (jonbob) [reply cime_config/config_grids.xml:4158]: @fdongyu - I do have another file for you to try, but anvil/chrysalis seem unusable right now. Are you set up to test on any other platforms?

Comment (fdongyu) [reply cime_config/config_grids.xml:4158]: @jonbob, thanks! Could you try Compy? All of my developments and tests are set up on Compy. I cannot access anvil/chrysalis and asked someone else to download the file for me.

Comment (jonbob) [reply cime_config/config_grids.xml:4158]: I have the file on compy, so that's easy:
/compyfs/inputdata/cpl/cpl6/map_EC30to60E2r2_to_r05_bilin.220728.nc

It's a typical bilinear remapping but uses a new(ish) extrapolation feature. I'm not sure it will be what you're looking for, but it's the only other thing I know of to try

Comment (fdongyu) [reply cime_config/config_grids.xml:4158]: @jonbob, this file results in weird remapped ssh (figure below). But I think this nearest neighbor remapping file you generated  (`/lcrc/group/e3sm/data/inputdata/cpl/cpl6/map_EC30to60E2r2_to_r05_neareststod.220728.nc` )
should be used because its only difference from my approach is that the land cells are not excluded from the remapping. This is OK because the ssh on the land cells will never be used. Only the ssh in the river outlet cells will be used, whose value comes from the nearest ocean cell.  

If it is OK, I would like to proceed with your file. My script could be an option in the future after more tests (maybe with the unified land/ocean mesh) in the ICoM group.  What do you think? 

remapped_ssh_bilinear

Comment (rljacob) [reply cime_config/config_grids.xml:4158]: > @rljacob as another possibility, what do you think of https://github.com/E3SM-Project/E3SM-init?

Yes that's another possibility.

Comment (jonbob) [reply cime_config/config_grids.xml:4158]: @fdongyu - that sounds good. I didn't have much faith in this last extrapolated mapping, so let's go forward with the ncremap neareststod approach. But please do keep your script ready in case we find issues with the single-point nearest neighbor

Comment (fdongyu) [reply cime_config/config_grids.xml]: @jonbob, thanks. The mapping file is now updated. The result does not show much difference compared with that using my previous mapping file.

**Review by jonbob — APPROVED** [review]
Approved based on testing and resolution of small issue

**Review by bishtgautam — APPROVED** [review]

**Review by hydrotian — APPROVED** [review]

**Review by liho745 — APPROVED** [review]

**Comment by jonbob** [comment]
@rljacob, @ambrad - are you OK with the resolution we came up with? I'd like to get this merged soon

**Comment by ambrad** [comment]
Hi @jonbob I think I'm on the reviewer list just b/c I made a comment at some point. You don't need to wait for my approval. I can't seem to take myself off the list.

**Comment by jonbob** [comment]
Thanks @ambrad -- sounds good

**Review by rljacob — APPROVED** [review]

**Comment by jonbob** [comment]
Passes:
* ERS.ne11_oQU240.WCYCL1850NS.chrysalis_intel
* SMS_D_Ld1.ne30pg2_r05_EC30to60E2r2.WCYCL1850.chrysalis_intel
* PEM_Ln9.ne30pg2_EC30to60E2r2.WCYCL1850.chrysalis_intel
with expected NML DIFFs

merged to next

**Comment by jonbob** [comment]
merged to master and expected NML DIFFs blessed, as well as new baselines requested for added test

**Pull request merged**
